agent/pkg/podman/podman.go

```diff
@@ -0,0 +1,481 @@
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/master/pkg/aproto"
+	"github.com/determined-ai/determined/master/pkg/cproto"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+)
+
+var podmanWrapperEntrypoint = path.Join(tasks.RunDir, tasks.SingularityEntrypointWrapperScript)
+
+const (
+	// TODO(DET-9111): Parameterize this by agent ID.
+	agentTmp         = "/tmp/determined/agent"
+	hostNetworking   = "host"
+	bridgeNetworking = "bridge"
+	envFileName      = "envfile"
+	archivesName     = "archives"
+)
+
+// PodmanContainer captures the state of a container.
+type PodmanContainer struct {
+	PID     int            `json:"pid"`
+	Req     cproto.RunSpec `json:"req"`
+	TmpDir  string         `json:"tmp_dir"`
+	Proc    *os.Process    `json:"-"`
+	Started atomic.Bool    `json:"started"`
+}
+
+// PodmanClient implements ContainerRuntime.
+type PodmanClient struct {
+	log        *logrus.Entry
+	opts       options.PodmanOptions
+	mu         sync.Mutex
+	wg         waitgroupx.Group
+	containers map[cproto.ID]*PodmanContainer
+}
+
+// New returns a new podman client, which launches and tracks containers.
+func New(opts options.PodmanOptions) (*PodmanClient, error) {
+	if err := os.RemoveAll(agentTmp); err != nil {
+		return nil, fmt.Errorf("removing agent tmp from previous runs: %w", err)
+	}
+
+	if err := os.MkdirAll(agentTmp, 0o700); err != nil {
+		return nil, fmt.Errorf("preparing agent tmp: %w", err)
+	}
+
+	return &PodmanClient{
+		log:        logrus.WithField("component", "podman"),
+		opts:       opts,
+		wg:         waitgroupx.WithContext(context.Background()),
+		containers: make(map[cproto.ID]*PodmanContainer),
+	}, nil
+}
+
+// Close the client, killing all running containers and removing our scratch space.
+func (s *PodmanClient) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Since we launch procs with exec.CommandContext under s.wg's context, this cleans them up.
+	s.wg.Close()
+
+	if err := os.RemoveAll(agentTmp); err != nil {
+		return fmt.Errorf("cleaning up agent tmp: %w", err)
+	}
+	return nil
+}
+
+func getPullCommand(req docker.PullImage, image string) (string, []string) {
+	// C.f. singularity where if req.ForcePull is set then a 'pull --force' is done.
+	// podman does not have this option, though it does have '--pull always' on the
+	// run command.
+	return "podman", []string{"pull", image}
+}
+
+// PullImage implements container.ContainerRuntime.
+func (s *PodmanClient) PullImage(
+	ctx context.Context,
+	req docker.PullImage,
+	p events.Publisher[docker.Event],
+) (err error) {
+	return cruntimes.PullImage(ctx, req, p, &s.wg, s.log, getPullCommand)
+}
+
+// CreateContainer implements container.ContainerRuntime.
+func (s *PodmanClient) CreateContainer(
+	ctx context.Context,
+	id cproto.ID,
```

Comment:
Who supplies this ID?

Reply:
It seems to come from the StartContainer request?